File: 438_Find_All_Anagrams_in_a_String.java
/*
https://leetcode.com/problems/find-all-anagrams-in-a-string/
Given a string s and a non-empty string p, find all the start indices of p's anagrams in s.
Strings consists of lowercase English letters only.

Sliding window over s of length of p, counting frequencies of all chars of s in window minus all chars in p. Also
track number of non-zero letters.
Time - O(n)
Space - O(1)
*/

public class Solution {
    public List<Integer> findAnagrams(String s, String p) {

        List<Integer> result = new LinkedList<Integer>();
        if (p.length() > s.length())
            return result;
        int[] window = new int[26];

        for (int i = 0; i < p.length(); ++i) {      // initialise window for length of p
            char c = p.charAt(i);
            --window[p.charAt(i) - 'a'];
            ++window[s.charAt(i) - 'a'];
        }

        int usedChars = 0;
        for (int i = 0; i < 26; ++i) {
            if (window[i] != 0)
                ++usedChars;
        }

        if (usedChars == 0)
            result.add(new Integer(0));

        for (int i = p.length(); i < s.length(); ++i) {
            ++window[s.charAt(i) - 'a'];
            if (window[s.charAt(i) - 'a'] == 0)
                --usedChars;
            if (window[s.charAt(i) - 'a'] == 1)
                ++usedChars;

            --window[s.charAt(i - p.length()) - 'a'];
            if (window[s.charAt(i - p.length()) - 'a'] == 0)
                --usedChars;
            if (window[s.charAt(i - p.length()) - 'a'] == -1)
                ++usedChars;

            if (usedChars == 0)
                result.add(new Integer(i - p.length() + 1));
            }

        return result;
    }
}

File: 438_Find_All_Anagrams_in_a_String.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/find-all-anagrams-in-a-string/
# Given a string s and a non-empty string p, find all the start indices of p's anagrams in s.
# Strings consists of lowercase English letters only and the length of both strings s and p will not be larger than 20,100.
# The order of output does not matter.

# Maintain a count of each letter in a sliding window in s, minus the counts of letters in p. Populate the dictionary
# with the intial counts, then slide along removing chars at the back and adding chars at the front. Add to result if
# dictionary is empty (all values zero).
# Time - O(n) len(s)
# Space - O(m) len(p)

from collections import defaultdict

class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        n = len(p)
        freq = defaultdict(int)     # map from char to net count of char in sliding window
        result = []

        if n > len(s):              # no anagrams are possible
            return result

        def update_freq(c, step):   # updates dictionary and deletes zero values
            freq[c] += step
            if freq[c] == 0:
                del freq[c]

        for c1, c2 in zip(p, s[:n]):    # populate initial window
            update_freq(c1, -1)
            update_freq(c2, 1)

        for i in range(len(s) - n):
            if not freq:
                result.append(i)
            update_freq(s[i], -1)       # remove char at back of window
            update_freq(s[i + n], 1)    # add char at front of window

        if not freq:
            result.append(len(s) - n)
        return result
